kolibri/auth/permissions.py

```diff
@@ -0,0 +1,349 @@
+"""
+This module defines the base classes for Kolibri's class-based Permissions system. Other apps can import the classes
+from this module in their own "permissions.py" module, and extend them, and then apply them to their own models.
+"""
+
+
+#####################################################################################################################
+### This first section consists of the base permissions classes: the ones likely to be imported and extended.     ###
+#####################################################################################################################
+
+class BasePermissions(object):
+    """
+    Base Permission class from which all other Permission classes should inherit.
+    By default, it denies permissions for everything. To be more permissive, override its methods.
+    Note that if you don't need to distinguish between create, update, and delete permissions, you can
+    simply override `_user_can_write_object` to define behavior for them in one place.
+    """
+
+    def user_can_create_object(self, user, obj):
+        """Returns True if this permission class grants <user> permission to create the provided <obj>.
+        Note that the object may not yet have been saved to the database (as this may be a pre-save check)."""
+        return self._user_can_write_object(user, obj)
+
+    def user_can_read_object(self, user, obj):
+        """Returns True if this permission class grants <user> permission to read the provided <obj>."""
+        return False
+
+    def user_can_update_object(self, user, obj):
+        """Returns True if this permission class grants <user> permission to update the provided <obj>."""
+        return self._user_can_write_object(user, obj)
+
+    def user_can_delete_object(self, user, obj):
+        """Returns True if this permission class grants <user> permission to delete the provided <obj>."""
+        return self._user_can_write_object(user, obj)
+
+    def _user_can_write_object(self, user, obj):
```

Comment:
Okay, do what you please. I think it's awkwardly unconventional to have a "private" method that must be overridden, but if it makes the implementation very simple and less error-prone then it's fine.

In my opinion, it makes sense to have a reasonable default (defer to `user_can_write_object`, and then require that only `user_can_write_object` be defined). In the exceptional case where each of the write permissions should be determined separately, the onus is on the implementer to do it "correctly" (i.e., override the right permissions methods and maybe cause `user_can_write_object` to raise an exception so that it's not inadvertently called). Perhaps we could also override [__getattr__](https://docs.python.org/2/reference/datamodel.html#object.__getattr__) to raise an exception when two mutually exclusive permissions methods are defined, if "Pythonic" defensive coding is your goal. There is some balance to be achieved between defensive coding and assuming the implementer is reasonably well-informed, and the latter can be aided with good documentation.